compiler/rustc_resolve/src/late/diagnostics.rs

```diff
@@ -744,15 +744,40 @@ impl<'a> LateResolutionVisitor<'a, '_, '_> {
                 }
             }
             (Res::Def(DefKind::Struct, def_id), _) if ns == ValueNS => {
-                if let Some((ctor_def, ctor_vis)) = self.r.struct_constructors.get(&def_id).cloned()
+                if let Some((ctor_def, ctor_vis, fields)) =
+                    self.r.struct_constructors.get(&def_id).cloned()
                 {
                     let accessible_ctor =
                         self.r.is_accessible_from(ctor_vis, self.parent_scope.module);
                     if is_expected(ctor_def) && !accessible_ctor {
-                        err.span_label(
-                            span,
-                            "constructor is not visible here due to private fields".to_string(),
-                        );
+                        let mut better_diag = false;
+                        if let PathSource::TupleStruct(_, ident_bindings) = &source {
+                            if ident_bindings.len() > 0 && fields.len() == ident_bindings.len() {
+                                let non_visible_spans: Vec<Span> = fields.iter().zip(ident_bindings.iter()).filter_map(|(vis, ident)| {
+                                    match self.r.is_accessible_from(*vis, self.parent_scope.module) {
+                                        true => None,
+                                        false => Some(ident.span),
+                                    }
+                                }).collect();
+                                // Extra check to be sure
+                                if non_visible_spans.len() > 0 {
+                                    let first_span = non_visible_spans[0];
+                                    err.set_span(non_visible_spans);
+                                    err.span_label(first_span, "private field");
+                                    err.set_primary_message(
+                                    "constructor is not visible here due to private fields".to_string(),
+                                    );
```

Comment:
this gives the weird format

   ^ ^
   |    message

@petrochenkov how do I get your suggestion:

   ^ ^ message

Reply:
I wrote `^ ^ message` from memory, so it probably won't look like that, but the goal is to use `span_labels` or `span_note` which accept multiple spans, and they will lay out the diagnostics in some way that is optimal from their point of view.

Reply:
ok this is using span_label so I think its about as optimal as it can get, but I will check in wg-diagnostics on zulip if there is a better way

Reply:
Maybe instead of `set_primary_message` use `span_note`? https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/struct.Diagnostic.html#method.span_note

Reply:
ok this is using span_label so I think its about as optimal as it can get, but I will check in wg-diagnostics on zulip if there is a better way

Reply:
@jyn514 span_note adds a new diagnostic with a top level message

Reply:
It seems to me that only using a single call to span label for each non-visible field would be enough to address the issue. That being said you can create a new MultiSpan, use push_label on it to have span messages to them in the subwindow and then a span_note call with that MultiSpan that will have a message. The API is a bit involved, but it wasn't originally intended for this.

Reply:
To clarify, the `set_primary_message` is not really intended to be used, so if we can improve things with a span_note, lets do that.

Reply:
The choice here is probably between `span_label` for the first private field and `span_labels` for all private fields.
`span_note` will look worse in the context of this specific diagnostic, and `set_primary_message` is not necessary, you certainly can construct the error with an immediate message here.

Reply:
I am not sure I understand, what do you mean "i can can construct the error with an immediate message here"?

Reply:
You can disregard this message entirely, I didn't look where the `err` was constructed.